### ISSUE TITLE
problem with DSM 7 (FritzOS 7.39)

### ISSUE THREAD

**Issue by pbuechel** [biocoder-frodo/FRITZBox-SynologyLogCenterDaemon#2]
After loging into Fritzbox it seems it's not possible to read the eventlog:

PHP Fatal error:  Uncaught PDOException: SQLSTATE[HY000] [14] unable to open database file in /volume1/Daten/Software/Backup/srv3/fritzbox-syslog-daemon/LogCenter.php:116
Stack trace:
#0 /volume1/Daten/Software/Backup/srv3/fritzbox-syslog-daemon/LogCenter.php(64): getLogCenterDb('/volume1/Daten/...', 'ddf-gw')
#1 /volume1/Daten/Software/Backup/srv3/fritzbox-syslog-daemon/fritzbox-syslog-daemon.php(49): getLogCenterTimeStamps('/volume1/Daten/...', 'ddf-gw')
#2 {main}
  thrown in /volume1/Daten/Software/Backup/srv3/fritzbox-syslog-daemon/LogCenter.php on line 116

**Comment by biocoder-frodo** [comment]
By coincidence, your PHP stack trace mentions issue number #1, did you start the script using sudo? The location looks a bit odd to me, which DSM version are you using?
Please try the updated script with the -d option, as shown here:

```
biocoder@diskstation:~$ php72 ./fritzbox-syslog-daemon.php -d -l=/var/services/homes/admin/logs/
Failed to open the database to get the record count of the logs table. Insufficient privileges or file not found?
Caught exception: SQLSTATE[HY000] [14] unable to open database file
int(-1)
biocoder@diskstation:~$ sudo php72 ./fritzbox-syslog-daemon.php -d -l=/var/services/homes/admin/logs/
int(400)
```

**Comment by pbuechel** [comment]
I'm running DSM 7.0.1-42218 Update 2. "Sudo" I have to try, but I'v to less time rights now.
I just ran a test with parameter -d but it seems there is no difference:

[script.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/8059409/script.log)
[output.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/8059410/output.log)
.

**Comment by pbuechel** [comment]
I tested with sudo and got the same errors on console. Sorry.
About the location: I put the stuff in a public place and not in the admin-home. Problem?

**Comment by biocoder-frodo** [comment]
With the updated [script](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/commit/e079b49569a828ca9afec911fcc836c0ab62dd9d) you should see the same error and int(-1) or no error and a positive or zero number of rows.

You can see [here](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/blob/e079b49569a828ca9afec911fcc836c0ab62dd9d/LogCenter.php#L107) that the database is expected to have a name according to 
`$dir = 'sqlite:'.$path.$host.'/SYNOSYSLOGDB_'.$host.'.DB';`

Note that the trailing path separator / needs to come from your configured value, so if your database is you where say that it is, you should add the / at the end of -l=/volume1..... argument.

Do you have an actual .DB file named `/volume1/Daten/Software/Backup/srv3/fritzbox-syslog-daemon/ddf-gw/SYNOSYSLOGDB_ddf-gw.DB` ??

If you followed the steps in the [wiki](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/wiki/Settings-on-your-Synology-NAS), I would expect your ` SYNOSYSLOGDB_ddf-gw.DB` to be present in `/var/services/homes/admin/logs/ `   and that its full path would be `/var/services/homes/admin/logs/ddf-gw/SYNOSYSLOGDB_ddf-gw.DB`

**Comment by pbuechel** [comment]
What I did:
- I moved everything, like in wiki, to /volume1/homes/admin/
- I added / to the -l=/volume1/homes...

I started the task. Attached you'll find the logs. There is no DB file anywhere. Where is this DB coming from? Does you script create it?

Starting with -d results in this:
```
Failed to open the database to get the record count of the logs table. Insufficient privileges or file not found?
Caught exception: SQLSTATE[HY000] [14] unable to open database file

```
Permissions are:
```
admin@srv3:~$ pwd
/var/services/homes/admin
admin@srv3:~$
admin@srv3:~$ ls -la
total 52
drwxrwxrwx+ 1 admin         users  382 Feb 15 06:57 .
drwxrwxrwx+ 1 root          root    94 Dec 12  2019 ..
-rwxrwxrwx+ 1 administrator users 3331 Feb 13 23:02 fritzbox-syslog-daemon.php
-rwxrwxrwx+ 1 administrator users  467 Feb 13 23:02 FritzLogCenterEvent.php
-rwxrwxrwx+ 1 administrator users 1123 Feb 13 23:02 FritzLogEvent.php
-rwxrwxrwx+ 1 administrator users 1023 Feb 13 23:02 FritzLog.php
-rwxrwxrwx+ 1 administrator users  980 Feb 13 23:02 FritzLuaLog.php
-rwxrwxrwx+ 1 administrator users 4858 Feb 13 23:02 FritzLuaSession.php
-rwxrwxrwx+ 1 administrator users 1063 Feb 13 23:02 LICENSE
-rwxrwxrwx+ 1 administrator users 2860 Feb 13 23:02 LogCenter.php
drwxrwxrwx+ 1 administrator users    0 Feb 15 06:57 logs
drwxrwxrwx+ 1 admin         users   44 Oct 19  2020 .Maildir
-rwxrwxrwx+ 1 administrator users 1755 Feb 13 23:02 README.md
-rwxrwxrwx+ 1 administrator users   13 Jan 29 10:25 stats.pwdfile
-rwxrwxrwx+ 1 administrator users  119 Feb 13 23:02 SysLogEvent.php
-rwxrwxrwx+ 1 administrator users 2871 Feb 13 23:02 UdpLog.php


```

[script.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/8066311/script.log)
[output.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/8066312/output.log)

**Comment by biocoder-frodo** [comment]
It's been a while, but you can give the latest version a try. This version can send a syslog packet that will create the database when Logcenter is running and listening on the port you specified. Start the script with the -i option and specify the root log folder for your databases,

**Comment by pbuechel** [comment]
Sorry, no difference:
[script.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9655188/script.log)
[output.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9655189/output.log)

**Comment by biocoder-frodo** [comment]
Ok, what is the output of the initialization command?

```
sudo php72 ./fritzbox-syslog-daemon.php -i -l=/volume2/homes/admin/logs/ -r=fritz.repeater -t=http --udp=517
pid =3840
array(5) {
  ["i"]=>
  bool(false)
  ["l"]=>
  string(26) "/volume2/homes/admin/logs/"
  ["r"]=>
  string(14) "fritz.repeater"
  ["t"]=>
  string(4) "http"
  ["udp"]=>
  string(3) "517"
}
Database should be created from syslog packet...
2 row(s) removed.
```

**Comment by pbuechel** [comment]
Here's the output:

```
administrator@srv3:/volume1/Daten/Software/Backup/LogCenterDaemon$ sudo php72 ./fritzbox-syslog-daemon.php -i -l=/volum
e1/Daten/Software/Backup/Logs/ -t=http -f=ddf-gw.fritz.box -u=stats
pid =16281
array(5) {
  ["i"]=>
  bool(false)
  ["l"]=>
  string(36) "/volume1/Daten/Software/Backup/Logs/"
  ["t"]=>
  string(4) "http"
  ["f"]=>
  string(16) "ddf-gw.fritz.box"
  ["u"]=>
  string(5) "stats"
}
Database should be created from syslog packet...
PHP Fatal error:  Uncaught PDOException: SQLSTATE[HY000] [14] unable to open database file in /volume1/Daten/Software/Backup/LogCenterDaemon/LogCenter.php:116
Stack trace:
#0 /volume1/Daten/Software/Backup/LogCenterDaemon/LogCenter.php(121): getLogCenterDb('/volume1/Daten/...', 'ddf-gw.fritz.bo...')
#1 /volume1/Daten/Software/Backup/LogCenterDaemon/fritzbox-syslog-daemon.php(43): removeDbInitMarker('/volume1/Daten/...', 'ddf-gw.fritz.bo...')
#2 {main}
  thrown in /volume1/Daten/Software/Backup/LogCenterDaemon/LogCenter.php on line 116

Fatal error: Uncaught PDOException: SQLSTATE[HY000] [14] unable to open database file in /volume1/Daten/Software/Backup/LogCenterDaemon/LogCenter.php:116
Stack trace:
#0 /volume1/Daten/Software/Backup/LogCenterDaemon/LogCenter.php(121): getLogCenterDb('/volume1/Daten/...', 'ddf-gw.fritz.bo...')
#1 /volume1/Daten/Software/Backup/LogCenterDaemon/fritzbox-syslog-daemon.php(43): removeDbInitMarker('/volume1/Daten/...', 'ddf-gw.fritz.bo...')
#2 {main}
  thrown in /volume1/Daten/Software/Backup/LogCenterDaemon/LogCenter.php on line 116
```

**Comment by pbuechel** [comment]
Maybe better in a file:
[new 1.txt](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9656089/new.1.txt)

**Comment by biocoder-frodo** [comment]
offtopic: If you select your text after you paste console output in the writing area, clicking the < > icon should mark the whole block as code :)

Ok, it looks like the packet was created succesfully, but the database step failed. If you open Logcenter now, do you have a ddf-gw.fritz.box section in the Logs pane of Logcenter? If Logcenter was running and accepted the message the script sent, you should see one row there. If that is the case, then the database is not where you expect it to be.

**Comment by pbuechel** [comment]
Yes, now I have a ddf-gw section. See pic. But what file I've to search for?
![Unbenannt](https://user-images.githubusercontent.com/98641867/192538820-1ecf98e5-e62c-489a-aa72-933c89f0f0f4.JPG)

**Comment by biocoder-frodo** [comment]
I'm guessing you have a logs folder in the default admin account. That's how it worked out of the box on my NAS. You can use File Station to search for a logs folder in a home directory of the, or one of the, admin accounts on your NAS. So browse the home(s) folder and look for files named SYNOSYSLOGDB*

![image](https://user-images.githubusercontent.com/3184378/192540101-b337facf-0e52-4e98-8996-228ef15bdb06.png)

**Comment by pbuechel** [comment]
No chance.
![Unbenannt](https://user-images.githubusercontent.com/98641867/192542008-8b4ba3b9-7224-43d4-8ef9-da62c7d76d3c.JPG)

**Comment by biocoder-frodo** [comment]
Mmh, on DSM 6 the search in homes finds my database files. It could be that the search hides files owned by root on your NAS. Please check the admin and administrator folders manually? I currently don't have immediate access to a DSM7 NAS, so I'm not sure where the database files would be if they are not inside a user profile.

@pbuechel : Below command should find your databases:
`sudo find / -iname "SYNOSYSLOG*.DB"`
or 
`sudo find / -iname "*ddf-gw*.DB"` ?

**Comment by pbuechel** [comment]
What I've found out:

When I use -i parameter a database SYNOSYSLOGDB_ARCH is created in a directory spcified in LogCenter-StorageDestinationForAchives.
When I specify this directory with the -l paramater the error is still the same.
A sudo find / -iname "*ddf-gw*.DB" doesn't find anything.

**Comment by biocoder-frodo** [comment]
> database SYNOSYSLOGDB_ARCH is created in a directory spcified in LogCenter-StorageDestinationForAchives.

Can you post a screenshot from File Station or a directory listing from the console?
The -l parameter assumes that DSM (6) places the databases for different log sources in subfolders in a common logs folder.

**Comment by pbuechel** [comment]
Hope I understood you correctly.
![Unbenannt2](https://user-images.githubusercontent.com/98641867/192558339-f3ed23ca-61e4-4dda-9f02-6e77fa9cfd13.JPG)
![Unbenannt1](https://user-images.githubusercontent.com/98641867/192558350-4a6b6afc-d7b1-4134-b374-b2ece575aa21.JPG)

**Comment by biocoder-frodo** [comment]
From the comment i made here https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/issues/2#issuecomment-1259514495, i anticipated that you would have a subfolder named ddf-gw.fritz.box inside your srv3 folder. Instead you have a FRITZSyslogDaemon folder. What is inside that folder?

**Comment by pbuechel** [comment]
LogFiles:
![Unbenannt](https://user-images.githubusercontent.com/98641867/192562490-95d952df-407b-4663-9998-a8695576cae9.JPG)

**Comment by biocoder-frodo** [comment]
I'm curious to see whether the database format in the those 3 *.DB* files is still the same as for DSM 6. 
If you want to scan them first, download 	sqlite-tools-win32-x86-3390300.zip (1.88 MiB) from [sqlite.org](https://www.sqlite.org/download.html) or the equivalent yet for your OS. Download the *.DB with File Station to your machine and open the database
with:
`sqlite3.exe SYNOSYSLOGDB__ARCH.DB`
E.g.:

```
sqlite3.exe SYNOSYSLOGDB_fritz.box.DB
SQLite version 3.39.3 2022-09-05 11:02:23
Enter ".help" for usage hints.
sqlite> .tables
logs
sqlite> select * from logs
```

You can hit Ctrl-C to close the SQLite client utility.
If you could upload the files, that would be great.

**Comment by pbuechel** [comment]
No problem:
[SYNOSYSLOGDB__ARCH.zip](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9657687/SYNOSYSLOGDB__ARCH.zip)

**Comment by pbuechel** [comment]
Now I'm off for today and tommorow. Thanks for your help and work so far.

**Comment by biocoder-frodo** [comment]
Same to you:)

I think the host is no longer a part of the database name, I need to modify the script quite a bit to support this changed behaviour..

```
sqlite3.exe pbuechel\SYNOSYSLOGDB__ARCH.DB
SQLite version 3.39.3 2022-09-05 11:02:23
Enter ".help" for usage hints.
sqlite> .tables
logs
sqlite> select * from logs;
1|ddf-gw.fritz.box|127.0.0.1|kern|info|info|06|1664288265|1664288265|+02:00|2022-09-27|16:17:45|SysLogDaemon|Database initialization message
sqlite>
```

**Comment by biocoder-frodo** [comment]
Give it a try after you update LogCenter.php and the main script from https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/tree/dsm-7

**Comment by pbuechel** [comment]
Now we are one steop further, I think. New error:
[script.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9672491/script.log)
[output.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9672493/output.log)

**Comment by biocoder-frodo** [comment]
To be sure about the code, could you post a response.json file from the latest commit to dsm-7 branch? If necessary, wipe the Event Log on your fritz.box first?

**Comment by pbuechel** [comment]
Sorry, I dont know what you talking of: "could you post a response.json file from the latest commit to dsm-7 branch"

**Comment by biocoder-frodo** [comment]
Please see https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/commit/75861081a171b6148e4b53b2d23c15fbb1528f10

With the added line, the latest response from the FRITZ!Box is saved to a file. I'd like to see the contents to see what is causing the error. It will contain the information from your Event Log, please review before sharing sensitive details.

**Comment by pbuechel** [comment]
I added the line an ran the script again but I cannot fina file named response.json anywhere ???
[output.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9676492/output.log)
[script.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9676495/script.log)

**Comment by biocoder-frodo** [comment]
I ran the script from my home folder, and the file was placed there. 
You can search for the file by entering `sudo find / -iname "response.json"`.
You can also replace `file_put_contents("response.json",$response);` with `echo $response . PHP_EOL;`.
In that case it will dump it straight to your output file.

**Comment by pbuechel** [comment]
Sorry, I didn't run the script interactively. Now I did and the output file was in the homedir of admin.
[response.zip](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9678035/response.zip)

**Comment by biocoder-frodo** [comment]
Please see https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/tree/dsm-7

**Comment by pbuechel** [comment]
Sorry:

administrator@srv3:~$ php72 /volume1/Daten/Software/Backup/LogCenterDaemon/fritzbox-syslog-daemon.php -l=/volume1/Daten
/Software/Backup/srv3/ -t=http -f=ddf-gw.fritz.box -u=stats -q -a -j
pid =27769
array(6) {
  ["l"]=>
  string(36) "/volume1/Daten/Software/Backup/srv3/"
  ["t"]=>
  string(4) "http"
  ["f"]=>
  string(16) "ddf-gw.fritz.box"
  ["u"]=>
  string(5) "stats"
  ["q"]=>
  bool(false)
  ["j"]=>
  bool(false)
}
PHP Fatal error:  Uncaught Error: Cannot use object of type stdClass as array in /volume1/Daten/Software/Backup/LogCenterDaemon/FritzLuaLog.php:15
Stack trace:
#0 /volume1/Daten/Software/Backup/LogCenterDaemon/FritzLuaLog.php(58): FritzLuaLog->parselogs(Object(stdClass), '')
#1 /volume1/Daten/Software/Backup/LogCenterDaemon/fritzbox-syslog-daemon.php(79): FritzLuaLog->getlogs('')
#2 {main}
  thrown in /volume1/Daten/Software/Backup/LogCenterDaemon/FritzLuaLog.php on line 15

Fatal error: Uncaught Error: Cannot use object of type stdClass as array in /volume1/Daten/Software/Backup/LogCenterDaemon/FritzLuaLog.php:15
Stack trace:
#0 /volume1/Daten/Software/Backup/LogCenterDaemon/FritzLuaLog.php(58): FritzLuaLog->parselogs(Object(stdClass), '')
#1 /volume1/Daten/Software/Backup/LogCenterDaemon/fritzbox-syslog-daemon.php(79): FritzLuaLog->getlogs('')
#2 {main}
  thrown in /volume1/Daten/Software/Backup/LogCenterDaemon/FritzLuaLog.php on line 15

**Comment by biocoder-frodo** [comment]
Now the script fails on my end when I enable the -j option and it works without it. I can start a rant about global variables, but I rushed it a bit this morning ;) Update here https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/tree/dsm-7

**Comment by pbuechel** [comment]
Perfect, now the LogCenter is filling. Thank you so much for your time and effort.

![Unbenannt](https://user-images.githubusercontent.com/98641867/193253773-87047bab-c626-4066-a8ec-ef7b8fb8dd5d.JPG)
[script.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9683436/script.log)
[output.log](https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/files/9683437/output.log)

**Comment by biocoder-frodo** [comment]
If you ever decide to buy another AVM device, you should update your code to the latest version as I found the -a option was not yet working properly for me. Since you have only one FRITZ!Box in your database it would not create an immediate problem. See https://github.com/biocoder-frodo/FRITZBox-SynologyLogCenterDaemon/pull/3

**Comment by pbuechel** [comment]
I own another AVM device, a repeater 2400 named DDF-REPEATER (see above pic).
But since he belongs to the mesh, it's log is replicated to ddf-gw by AVM. 
So where is my problem?

**Comment by biocoder-frodo** [comment]
If you were to invoke a second copy of the script for the DDF-REPEATER specifically, the script would fetch all entries from the log table when trying to determine new log rows from the device and would get confused. It would retrieve logs from two devices if I did not change the select queries.
![image](https://user-images.githubusercontent.com/3184378/193408533-b113058b-eb0b-4b1a-b50a-471c3df0a611.png)
![image](https://user-images.githubusercontent.com/3184378/193408612-644a265b-92f8-4a57-aa36-5cc56c56fac5.png)

**Comment by pbuechel** [comment]
So if I start a second php instance for the repeater, the first instance will filter out the messages from the FRITZ!box log regarding the repeater? 
just to be sure, currently the FRITZ!box of contains logs from the FRITZ!box AND repeater.

**Comment by biocoder-frodo** [comment]
Your FRITZ!box logs some messages from the repeater because its in mesh mode. Those messages will not be filtered out. There is a filter option, but I don't use that on my 7583.
If you start a second instance for the repeater, the script has to be sure the log rows came from the repeater. In your NAS you have only one database table, and therefore the hostname has to be in the select query for either the router and the repeater.

**Comment by biocoder-frodo** [comment]
I'm going to close this issue as I started using FRITZ!OS 7.50 on my 7583. My Fritz!Repeater 3000 is still running on FRITZ!OS 7.29 and does not need the -j option yet, but the 7583 now needs it. Everything was fine after I reconfigured the Task Scheduler for the 7583.

From the last logs:
FRITZ!OS updated to 7.50. The previously used version was 7.31. The update was performed using the update search in the user interface.
